src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/HaplotypeCallerEngine.java

```diff
@@ -0,0 +1,861 @@
+            genotypes.add( new GenotypeBuilder(sample.getKey()).alleles(noCall).PL(genotypeLikelihoods).make() );
+        }
+
+        final List<Allele> alleles = Arrays.asList(FAKE_REF_ALLELE , FAKE_ALT_ALLELE);
+        final double isActiveProb;
+
+        if (genotypes.size() == 1) {
+            // Faster implementation avoiding the costly and over complicated Exact AFCalculator machinery:
+            // This is the case when doing GVCF output.
+            isActiveProb = activeRegionEvaluationGenotyperEngine.calculateSingleSampleRefVsAnyActiveStateProfileValue(genotypes.get(0).getLikelihoods().getAsVector());
+        } else {
+            final VariantCallContext vcOut = activeRegionEvaluationGenotyperEngine.calculateGenotypes(new VariantContextBuilder("HCisActive!", context.getContig(), context.getLocation().getStart(), context.getLocation().getEnd(), alleles).genotypes(genotypes).make(), GenotypeLikelihoodsCalculationModel.Model.SNP, readsHeader);
+            isActiveProb = vcOut == null ? 0.0 : QualityUtils.qualToProb(vcOut.getPhredScaledQual());
+        }
+        return new ActivityProfileState(ref.getInterval(), isActiveProb, averageHQSoftClips.mean() > AVERAGE_HQ_SOFTCLIPS_HQ_BASES_THRESHOLD ? ActivityProfileState.Type.HIGH_QUALITY_SOFT_CLIPS : ActivityProfileState.Type.NONE, averageHQSoftClips.mean() );
+    }
+
+    /**
+     * Generate variant calls for an assembly region
+     *
+     * @param region region to assemble and perform variant calling on
+     * @param features Features overlapping the assembly region
+     * @return List of variants discovered in the region (may be empty)
+     */
+    public List<VariantContext> callRegion( final AssemblyRegion region, final FeatureContext features ) {
+        if ( hcArgs.justDetermineActiveRegions ) {
+            // we're benchmarking ART and/or the active region determination code in the HC, just leave without doing any work
+            return NO_CALLS;
+        }
+
+        if ( hcArgs.sampleNameToUse != null ) {
+            removeReadsFromAllSamplesExcept(hcArgs.sampleNameToUse, region);
+        }
+
+        if( ! region.isActive() ) {
+            // Not active so nothing to do!
+            return referenceModelForNoVariation(region, true);
+        }
+
+        final List<VariantContext> givenAlleles = new ArrayList<>();
+        if ( hcArgs.genotypingOutputMode == GenotypingOutputMode.GENOTYPE_GIVEN_ALLELES ) {
+            for ( final VariantContext vc : features.getValues(hcArgs.alleles) ) {
+                if ( vc.isNotFiltered() ) {
+                    givenAlleles.add(vc); // do something with these VCs during GGA mode
+                }
+            }
+
+            // No alleles found in this region so nothing to do!
+            if ( givenAlleles.isEmpty() ) {
+                return referenceModelForNoVariation(region, true);
+            }
+        } else if( region.size() == 0 ) {
```

Comment:
it seems that this check `if( region.size() == 0 ){ return referenceModelForNoVariation(region, true); }` could me moved up to before the declaration of `givenAlleles`. It would make the code cleaner i think